ibis/common/patterns.py

```diff
@@ -0,0 +1,1196 @@
+from __future__ import annotations
+
+import inspect
+import math
+import numbers
+from abc import ABC, abstractmethod
+from collections.abc import Callable, Hashable, Mapping, Sequence
+from enum import Enum
+from inspect import Parameter
+from itertools import chain, zip_longest
+from typing import Any as AnyType
+from typing import (
+    Generic,  # noqa: F401
+    Literal,
+    Optional,
+    Tuple,
+    TypeVar,
+    Union,
+)
+
+from typing_extensions import Annotated, Self, get_args, get_origin
+
+from ibis.common.collections import RewindableIterator, frozendict
+from ibis.common.dispatch import lazy_singledispatch
+from ibis.common.typing import get_bound_typevars, get_type_params
+from ibis.util import is_iterable, promote_tuple
+
+try:
+    from types import UnionType
+except ImportError:
+    UnionType = object()
+
+
+T = TypeVar("T")
+
+
+class CoercionError(Exception):
+    ...
+
+
+class Coercible(ABC):
+    """Protocol for defining coercible types.
+
+    Coercible types define a special ``__coerce__`` method that accepts an object
+    with an instance of the type. Used in conjunction with the ``coerced_to``
+    pattern to coerce arguments to a specific type.
+    """
+
+    __slots__ = ()
+
+    @classmethod
+    @abstractmethod
+    def __coerce__(cls, value: Any, **kwargs: Any) -> Self:
+        ...
+
+
+class ValidationError(Exception):
+    ...
+
+
+class Validator(ABC):
+    __slots__ = ()
+
+    @abstractmethod
+    def validate(self, value, context):
+        ...
+
+
+class MatchError(Exception):
+    ...
+
+
+class NoMatch:
+    """Sentinel value for when a pattern doesn't match."""
```

Comment:
Minor minor nit, but could you use an enum here instead? This pattern plays better with type checkers, and makes it clearer that `NoMatch` is a singleton *value* rather than a singleton type (i.e. consumers should check for `is NoMatch` rather than `isinstance(x, NoMatch)`). Not a big deal though.

```python
class NoMatchType(enum.Enum):
    NoMatch = "NoMatch"

NoMatch = NoMatchType.NoMatch
```

This way if we ever wanted to type annotate the `match` methods below, their signature could be:

```python
def match(self, value: Value) -> Value | NoMatchType:
    ...
```

See https://peps.python.org/pep-0484/#support-for-singleton-types-in-unions for reference.

Reply:
Sounds good to me, but I'm deferring it to a follow-up if you don't mind.